straxen/plugins/position_reconstruction.py

```diff
@@ -203,3 +204,95 @@ def compute(self, peaks):
         for xy in ('x', 'y'):
             result[xy] = peaks[f'{xy}_{algorithm}']
         return result
+
+    
+    
+@export
+@strax.takes_config(
+    strax.Option('recon_alg_included', help='The list of all reconstruction algorithm considered.',
+                 default=('_mlp','_gcn','_cnn'),
+                )
+)
+class S2ReconPosDiff(strax.Plugin):
+    '''
+    Plugin that provides position reconstruction difference for S2s in events, see note: 
+    https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:shengchao:sr0:reconstruction_quality
+    '''
+    
+    __version__ = '0.0.2'
+    parallel = True
+    depends_on = 'event_basics'
+    provides = 's2_recon_pos_diff'
+    
+    def infer_dtype(self):
+        dtype =[
+        ('s2_recon_avg_x' ,np.float32,
+         'Mean value of x for main S2'),
+        ('alt_s2_recon_avg_x' ,np.float32,
+         'Mean value of x for alternatice S2'),
+        ('s2_recon_avg_y' ,np.float32,
+         'Mean value of y for main S2'),
+        ('alt_s2_recon_avg_y' ,np.float32,
+         'Mean value of y for alternatice S2'),
+        ('s2_recon_pos_diff' ,np.float32,
+         'Reconstructed position difference for main S2'),
+        ('alt_s2_recon_pos_diff' ,np.float32,
+         'Reconstructed position difference for alternative S2'),
+
+    ]
+        dtype += strax.time_fields
+        return dtype
+    
+    def compute(self,events):
+        
+        result = np.zeros(len(events), dtype=self.dtype)
+        result['time'] = events['time']
+        result['endtime'] = strax.endtime(events)
+        # Computing position difference
+        self.compute_pos_diff(events, result)
+        return result  
+
+    def cal_avg_and_std(self, values, axis=1):
+        average = np.average(values,axis=axis)
+        variance = np.var(values, axis=axis)
+        return (average, np.sqrt(variance))
+
+    def eval_recon(self, data, name,cur_s2_bool):
+        alg_list=self.config['recon_alg_included']
+        name_x_list=[]
+        name_y_list=[]
+        for alg in alg_list:
+            name_x_list.append(name+'_x'+alg)
+            name_y_list.append(name+'_y'+alg)
+        x_avg,x_std= self.cal_avg_and_std(np.array(data[name_x_list][cur_s2_bool].tolist())) #lazy fix for data types 
+        y_avg,y_std= self.cal_avg_and_std(np.array(data[name_y_list][cur_s2_bool].tolist()))

```

Comment:
Those lines I do not understand `data` aka `events` should be already an array is not it. I do not understand why you transform it here back and forth.

Reply:
np.average or np.mean does not use np.array with filed name as an input. After this operation, the filed name was neglected which it gets around the error.

Reply:
`np.mean(some_structure_array['some_field_name'])` should work.

Reply:
@WenzDaniel , if I understand the problem correctly, this is the issue (there are multiple fields):
```python
>>> import numpy as np
>>> a= np.ones(2, dtype=[('bla', float), ('foo', float)])
>>> a['foo'] = [2, 2]
>>> a
array([(1., 2.), (1., 2.)], dtype=[('bla', '<f8'), ('foo', '<f8')])
>>> np.mean(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 5, in mean
  File "/home/joran/miniconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py", line 3372, in mean
    return _methods._mean(a, axis=axis, dtype=dtype,
  File "/home/joran/miniconda3/lib/python3.9/site-packages/numpy/core/_methods.py", line 160, in _mean
    ret = umr_sum(arr, axis, dtype, out, keepdims)
TypeError: cannot perform reduce with flexible type
>>> np.mean(a.tolist())
1.5
```